ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/ModelConformanceValidator.scala

```diff
@@ -24,116 +19,125 @@ import com.daml.ledger.participant.state.kvutils.{Conversions, Err}
-        // not included in the inputs.
-        lazy val knownKeys: Map[DamlContractKey, Value.ContractId] =
-          commitContext.collectInputs {
-            case (key, Some(value))
-                if value.getContractState.hasContractKey
-                  && contractIsActive(transactionEntry, value.getContractState) =>
-              value.getContractState.getContractKey -> Conversions
-                .stateKeyToContractId(key)
-          }
+        val validationResult = engine.validate(
+          transactionEntry.submitters.toSet,
+          SubmittedTransaction(transactionEntry.transaction),
+          transactionEntry.ledgerEffectiveTime,
+          commitContext.participantId,
+          transactionEntry.submissionTime,
+          transactionEntry.submissionSeed,
+        )
+
+        for {
+          stepResult <- consumeValidationResult(
+            validationResult,
+            transactionEntry,
+            commitContext,
+            rejections,
+          )
+          finalStepResult <- validateCausalMonotonicity(stepResult, commitContext, rejections)
+        } yield finalStepResult
+      })
+  }
 
-        try {
-          engine
-            .validate(
-              transactionEntry.submitters.toSet,
-              SubmittedTransaction(transactionEntry.transaction),
-              transactionEntry.ledgerEffectiveTime,
-              commitContext.participantId,
-              transactionEntry.submissionTime,
-              transactionEntry.submissionSeed,
-            )
-            .consume(
-              lookupContract(transactionEntry, commitContext),
-              lookupPackage(commitContext),
-              lookupKey(commitContext, knownKeys),
-            )
-            .fold(
-              err =>
-                rejections.buildRejectionStep(
-                  transactionEntry,
-                  rejectionReasonForValidationError(err),
-                  commitContext.recordTime,
-                ),
-              _ => StepContinue[DamlTransactionEntrySummary](transactionEntry),
-            )
-        } catch {
-          case err: Err.MissingInputState =>
-            logger.warn(
-              "Model conformance validation failed due to a missing input state (most likely due to invalid state on the participant)."
-            )
+  private def consumeValidationResult(
+      validationResult: Result[Unit],
+      transactionEntry: DamlTransactionEntrySummary,
+      commitContext: CommitContext,
+      rejections: Rejections,
+  )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
+    try {
+      val stepResult = for {
+        contractKeyInputs <- transactionEntry.transaction.contractKeyInputs.left
+          .map(rejectionForKeyInputError(transactionEntry, commitContext.recordTime, rejections))
+        _ <- validationResult
+          .consume(
+            lookupContract(commitContext),
+            lookupPackage(commitContext),
```

Comment:
Not all of them. Errors that originate from the current ledger state i.e. `Err.MissingInputState` and `Err.DecodeError` end up with `Inconsistent` (see https://github.com/digital-asset/daml/pull/10066#discussion_r672076310). I'm actually not sure about `Err.DecodeError`. I wanted to preserve the current `lookupPackage` behavior and ignored it. Your comment made me realize that it was broken from the very beginning. There was no catch clause for `Err.DecodeError`. I changed it to a catch-all `Err` block and added more tests. Please check if it makes sense.

Reply:
I’m not quite sure Inconsistent makes sense.
1. For `DecodeError`, I think we’re really talking about an internal error. We’re not failing to decode what the participant send us, we’re failing to decode our own state. That should never happen and imho we shouldn’t report it as Inconsistent.
2. A missing package is similar to the missing contract case discussed in the other comment thread. However, there is no committer pruning of packages. So I think this can really only arise from a malicious/buggy participant. I think putting it under Inconsistent is fine since it’s not technically model conformance (because it isn’t a property of the transaction itself) but maybe at least add a comment that this can only arise from a malicous submitter.
3. For missing contracts, I think Inconsistent is probably the best choice.

Reply:
1. We can also not catch it at all, as it was previously. @SamirTalwar-DA told me that it's possible if we turn off the package validation in the Engine, and suggested a new rejection reason, which we probably don't want to introduce now.
2. Will add it tomorrow, thanks!

Reply:
1) `DecodeError`: I agree that it should not be reported as inconsistent but rather as disputed. From the documentation of disputed:
> This means that the underlying ledger and its validation logic
    considered the transaction potentially invalid. This can be due to a bug
    in the submission or validation logic, or due to malicious behaviour.

2) I agree with @cocreature's assessment.
3) I agree with @cocreature's assessment.

Reply:
Another note regarding 1): In fact, we already have an appropriate exception that could be thrown, bot be more explicit about the specific case: `Err.ArchiveDecodingFailed`.

Reply:
Done, thanks